calico_kubernetes/calico_kubernetes.py

```diff
@@ -186,62 +228,104 @@ def _container_add(self, pid, ip, interface, hostname, orchestrator_id):
+
+        logger.info("Setting mac address %s to endpoint %s", ep.mac, ep.name)
         self._datastore_client.set_endpoint(ep)
 
         # Let the caller know what endpoint was created.
         return ep
 
-    def _container_remove(self, hostname, orchestrator_id):
+    def _assign_container_ip(self):
+        """
+        Assign IPAddress either with the assigned docker IPAddress or utilize
+        calico IPAM.
+
+        The option to utilize IPAM is indicated by the environment variable
+        "CALICO_IPAM".
+        True indicates to utilize Calico's auto_assign IPAM policy.
+        False indicate to utilize the docker assigned IPAddress
+
+        :return list containing the IPAddress which has been assigned
+        """
+        if CALICO_IPAM == 'true':
+            # Assign ip address through IPAM Client
+            logger.info("Using Calico IPAM")
+            try:
+                ip_list, ipv6_addrs = self._datastore_client.auto_assign_ips(
+                    1, 0, self.docker_id, None)
+                logger.debug("ip_list is %s; ipv6_addrs is %s", ip_list, ipv6_addrs)
+                assert not ipv6_addrs
+            except RuntimeError as err:
+                logger.error("Cannot auto assign IPAddress: %s", err.message)
+                sys.exit(1)
+
+        else:
+            logger.info("Using docker assigned IP address")
+            ip_list = [self._read_docker_ip()]
+            try:
+                self._datastore_client.assign_ip(ip_list[0], str(self.docker_id), None)
+            except (ValueError, RuntimeError, AlreadyAssignedError):
+                logger.exception("Cannot assign IPAddress %s", ip_list[0])
+                sys.exit(1)
+
+        return ip_list
+
+    def _container_remove(self):
         """
         Remove the indicated container on this host from Calico networking
         """
         # Find the endpoint ID. We need this to find any ACL rules
         try:
             endpoint = self._datastore_client.get_endpoint(
-                hostname=hostname,
-                orchestrator_id=orchestrator_id,
+                hostname=HOSTNAME,
+                orchestrator_id=ORCHESTRATOR_ID,
                 workload_id=self.docker_id
             )
         except KeyError:
-            logger.error("Container %s doesn't contain any endpoints", self.docker_id)
+            logger.exception("Container %s doesn't contain any endpoints", self.docker_id)
             sys.exit(1)
 
         # Remove any IP address assignments that this endpoint has
+        ip_set = set()
         for net in endpoint.ipv4_nets | endpoint.ipv6_nets:
             assert(net.size == 1)
-            self._datastore_client.unassign_address(None, net.ip)
+            ip_set.add(net.ip)
 
-        # Remove the endpoint
-        netns.remove_veth(endpoint.name)
+        logger.info("Removing IP addresses %s from endpoint %s", ip_set, endpoint.name)
+        self._datastore_client.release_ips(ip_set)
```

Comment:
You could clean this up quite a bit by removing the assert(net.size==1) above (which doesn't buy us much).

```
# Release all IPs on this endpoint.
ip_set = set(endpoint.ipv4_nets | endpoint.ipv6_nets)
self._datastore_client.release_ips(ip_set)
```